### ISSUE TITLE
performance when opening many sources

### ISSUE THREAD

**Issue by martinschorb** [bigdataviewer/bigdataviewer-playground#138]
Hi,

I am loading ~100 sources and at some point the performance starts to drop significantly (5s to load a single 2kx2k 2d image).

Is this normal? I think file IO should be OK. I tred to copy from the same source directory in parallel and that was fast.

**Comment by NicoKiaru** [comment]
By loading, do you mean showing all sources together in a bdv window or just opening them in the 'tree view' ?

**Comment by martinschorb** [comment]
Hi,

just putting them in the tree.

In fact, I think it just started to completely freeze while doing so.

**Comment by martinschorb** [comment]
In total it is 260 individual XML/H5 pairs. Most of them are rather small (6MB) single tiles.

**Comment by NicoKiaru** [comment]
ok, so it's not related to the size of the image - but rather to the way this tree view is made and updated. Everything's up to date on your fiji ?  because the version ~ 1 month ago was not dealing well with a big amount of sources

**Comment by martinschorb** [comment]
The UI now is extremely lacking behind. Even scrolling the list takes seconds to refresh. Utilizes one full core to 100% and 3.5GB of RAM.

**Comment by martinschorb** [comment]
no updates shown. Have there been any in the last 2 weeks?

**Comment by NicoKiaru** [comment]
No that's lagging even with the update. I can reproduce this easily

**Comment by NicoKiaru** [comment]
ok, as a workaround there are ways to bundle many xml/hdf5  together in a single one but for the moment this is difficult to solve.

**Comment by tischi** [comment]
@NicoKiaru do you already know whether it is the `TreeView` or something else, e.g. inside BDV, that causes the slow down?

**Comment by NicoKiaru** [comment]
I checked and yes it's the treeview update called each time a source is append to it:

https://github.com/bigdataviewer/bigdataviewer-playground/blob/4777bc821b1f67eaa6fde41ac73e38d6b0cdb457/src/main/java/sc/fiji/bdvpg/scijava/services/SourceAndConverterService.java#L197

https://github.com/bigdataviewer/bigdataviewer-playground/blob/4777bc821b1f67eaa6fde41ac73e38d6b0cdb457/src/main/java/sc/fiji/bdvpg/scijava/services/SourceAndConverterService.java#L183


I don't know exactly how to solve this. Having the ui modular, fast, and redrawn when needed but not too much is not simple.

**Comment by tischi** [comment]
Maybe there should be an option for the user to load many sources in one go (maybe there is already?) and then you `ui.update(List<sac>)` all in one go?


...and is `ui.update()` `synchronized`? otherwise you may run into concurrency issues, maybe explaining the freezing.

**Comment by NicoKiaru** [comment]
The update on sources list may be the way to go. I don't know if this will be enough to resolve the performance issue but that's something needed I agree. 

Regarding concurrency issues - which are independent of this performance issue I believe:

Using synchronized, I can try, but I don't know if it's a method which should be synchronized or rather that we should make a synchronized block with the `treemodel` object as a lock, each time it is modified. Also the Swing thread(s?) updates the UI in a separate thread than bdv-playground.  I'm not sure how to resolve all issues but I noticed that some calls need to be done from the Swing thread by using `SwingUtilities.invokeLater(() -> doStuff())`. You'll find some calls like this in the repo if you search

**Comment by tischi** [comment]
I had another idea: inside the `ui.update(sac)` method: you just add the sac to a List "toBeUpdated" and trigger a countdown (say 100ms). If within those 100ms the `ui.update(sac)` the method is _not_ called another time, you update the Tree UI. If it _is_ called another time you add the next sac also to the list and reset the countdown. Like this, sacs that are added fast after each other would only cause one UI update.

**Comment by tischi** [comment]
...commenting on my own comments above: this of course only makes sense if the issue only occurs when **adding several sources at a time**. But reading @martinschorb's issue again I guess we also have a problem **just adding one source** to many, right? This sounds like this issue: https://stackoverflow.com/questions/2591735/swing-what-to-do-when-a-jtree-update-takes-too-long-and-freezes-other-gui-eleme

...and regarding the synchronisation: good points! I am not sure how to handle this perfectly either. 
- Invoking stuff with `SwingUtilities.invokeLater()` often helped freezing UI issues for me as well. 
- I think it will not hurt to synchronize the update method anyway, maybe one could become more sophisticated if needed. 

but as said all if this is kind of obsolete, if the issue also occurs just adding one source.

In general I think considering an alternative source UI to the Tree might be another idea. I think for end-users it may be a bit hard to understand the Tree anyway. Could we get away with (few) drop-downs and few buttons for the sources, as in MoBIE? Maybe this would be more performant than the tree UI?

**Comment by NicoKiaru** [comment]
Turns out : a stupid `System.out.println` was causing a very very important perf loss.

I can now open a 100 SpimData in the ui in a few seconds. Not terrible but that's way way better than before. Also I think I may have solved the concurrency issue in the tree viewer + the selection of sources which also had a performance issue (the rectangle drawing was called using paint instead of repaint).

**Comment by NicoKiaru** [comment]
I've added some additional perf improvement (calling the reload model  if necessary once every 100 ms) a bit like what you proposed @tischer . FIngers crossed it's not causing another issue. With that I can open 500 xml/hdf5 file in 17s. Let's see if that suits you @martinschorb . I'll try to push to the update site tomorrow evening.

**Comment by martinschorb** [comment]
Hi,

for me, the update site has not changed yet. Is it already there, so I can test a bit more?

**Comment by NicoKiaru** [comment]
Yep sorry, too much stuff to do. I'll try to push before 13h

**Comment by NicoKiaru** [comment]
OK, done @martinschorb try it and let me know. Side note : in theory you can drag and drop xml files in the tree view, but I did not test again this functionality.

+ I had to uninstall bdv-playground , restart fiji, then reinstall bdv-playground in order to see the update. I hope it was well updated

**Comment by martinschorb** [comment]
Hi,

This helped!
Now it fills the list like a charm...
